webapp/src/lib/encryption.ts

```diff
@@ -0,0 +1,194 @@
+	// private key with the Crypto Subtle API by just concatenating a prefix to
+	// the raw private key bytes (`#xx` represent byte lengths):
+	//
+	// 30 #2e             - `PrivateKeyInfo`[2] constructed sequence
+	//   02 #01 00        - `version` value of 0
+	//   30 #05           - `AlgorithmIdentifier`[3] constructed sequence
+	//     06 #03 2b656e  - The algorithm object identifier for x25519[4]
+	//   04 #22           - `PrivateKey` octet string
+	//     04 #20 <KEY>   - For x25519, the BER encoded 32-byte octet string
+	//
+	// [1]: <https://en.wikipedia.org/wiki/X.690>
+	// [2]: <https://datatracker.ietf.org/doc/html/rfc5208#section-5>
+	// [3]: <https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.1.2>
+	// [4]: <https://datatracker.ietf.org/doc/html/rfc8410#section-3>
+	const pkcs8 = ethers.getBytes(
+		ethers.concat(["0x302e020100300506032b656e04220420", secret]),
+	);
+	const privateKey = await crypto.subtle.importKey(
+		"pkcs8",
+		pkcs8,
+		{ name: "X25519" },
+		false,
+		["deriveBits"],
+	);
+
+	// Unfortunately, Crypto Subtle API does not have a good way to derive an
+	// X25519 public key from a private one - exporting as `jwk` from a PKCS #8
+	// imported private key does not work across browsers. Use `@noble/curves`
+	// to compute the raw public key instead.
+	const publicKeyRaw = x25519.getPublicKey(ethers.getBytes(secret));
+	const publicKey = await importPublicKey(publicKeyRaw);
+
+	return { publicKey, privateKey };
+}
+
+/**
+ * Import a raw x25519 public key to a Crypto Subtle key.
+ */
+async function importPublicKey(raw: BytesLike): Promise<CryptoKey> {
+	if (ethers.dataLength(raw) !== 32) {
+		throw new Error("invalid raw x25519 public key");
+	}
+	return await crypto.subtle.importKey(
+		"raw",
+		ethers.getBytes(raw),
+		{ name: "X25519" },
+		true,
+		[],
+	);
+}
+
+/**
+ * Exports a public key to its raw bytes.
+ */
+async function exportPublicKey({
+	publicKey,
+}: Pick<CryptoKeyPair, "publicKey">): Promise<Hex> {
+	const raw = await crypto.subtle.exportKey("raw", publicKey);
+	return ethers.hexlify(new Uint8Array(raw));
+}
+
+/**
+ * RLP-encode a Safe transaction for encryption.
+ */
+function rlpEncodeSafeTransaction(safeTx: SafeTransaction): Uint8Array {
+	const n = ethers.toBeArray;
+	return ethers.getBytes(
+		ethers.encodeRlp([
+			safeTx.to,
+			n(safeTx.value),
+			safeTx.data,
+			n(safeTx.operation),
+			n(safeTx.safeTxGas),
+			n(safeTx.baseGas),
+			n(safeTx.gasPrice),
+			safeTx.gasToken,
+			safeTx.refundReceiver,
+		]),
+	);
+}
+
+/**
+ * RLP-decode a Safe transaction for decryption.
+ */
+function rlpDecodeSafeTransaction(data: BytesLike): SafeTransaction {
+	const decoded = ethers.decodeRlp(data);
+	if (
+		!Array.isArray(decoded) ||
+		decoded.length !== 9 ||
+		decoded.some((field) => typeof field !== "string")
+	) {
+		throw new Error("invalid Safe transaction RLP encoding");
+	}
+	const fields = decoded as string[];
+
+	const a = ethers.getAddress;
```

Comment:
personal opinion (can be ignore): for me this does not make it easier to read the code. I would potentially have rather defined functions that are not exported and give them proper names (i.e. `parseFieldToQuantity` and `parseFieldToOperation`)

Reply:
Its a bad habit from my functional programming days - they love short 1-letter identifiers. The argument is that the type holds enough information, and your editor should be able to show the information.

Given the type-erasure of "RLP fields" in `ethers` and JS in general, we should be more explicit here.